### PR TITLE
Add kv_nonpad_seqlen input to Attention

### PR DESCRIPTION
**Pull request by yuanyao-nv**
### Description

To accompany the [TensorScatter-24](https://github.com/onnx/onnx/pull/7114) op for managing in-place KV cache update, this PR makes the following changes to the Attention op:
- Add `nonpad_kv_seqlen` to indicate the number of valid (nonpadded) tokens in the K and V inputs when the K and V inputs are the entire cache tensors (where the number of valid tokens can potentially make up only a small proportion of the cache tensors). The `nonpad_kv_seqlen` input would provided optimization opportunities for backends to skip the unnecessary computation on the padding tokens.
- Allow the kv_seqlen dimension (-1 dimension) of `attn_mask` input to be shorter than K and V. The missing portion will be assumed to be -inf. The length should still be larger than the max value in `nonpad_kv_seqlen`.

Also, allow `attn_mask` and `is_causal` to be present at the same time. This would allow for easier export of HF models later.

### CODE REVIEW

File: onnx/defs/nn/defs.cc [hunk 1]
```diff
@@ -3447,6 +3449,13 @@ ONNX_OPERATOR_SET_SCHEMA(
             "past state cache for value with shape `(batch_size, kv_num_heads, past_sequence_length, v_head_size)`",
             "T2",
             OpSchema::Optional)
+        .Input(
+            6,
+            "kv_nonpad_seqlen",
```
Comment (gramalingam) [inline onnx/defs/nn/defs.cc]: Another candidate for the name is "dynamic_kv_seqlen" ... or maybe "nonpadded_kv_seqlen" would be better.

File: onnx/defs/nn/defs.cc [hunk 2]
```diff
@@ -3447,6 +3449,13 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+            "A vector of integers of shape `(batch_size,)` that indicates the number of valid (ie, non-padded)"
+            "tokens in each sample. A padding mask can be derived from this.",
```
Comment (gramalingam) [inline onnx/defs/nn/defs.cc]: nit: Clarify that this cannot be used simultaneously with KV cache?

File: onnx/defs/nn/defs.cc [hunk 3]
```diff
@@ -3330,6 +3330,8 @@ Attention bias to be added is calculated based on `attn_mask` input and `is_caus
 1) If `is_causal` is set to `1`, the attention masking is a lower triangular matrix when the mask is a square matrix. The attention masking has the form of the upper left causal bias due to the alignment.
 2) `attn_mask`: A boolean mask where a value of `True` indicates that the element should take part in attention or a float mask of the same type as query, key, value that is added to the attention score.
 
+Optionally, `kv_nonpad_seqlen` can be provided to indicate the number of valid tokens in each sample.
```
Comment (gramalingam) [inline onnx/defs/nn/defs.cc]: Please add a description of how this handled. Clarify that it is of shape `(batchsize,)` and that it is used only to construct a corresponding causal-mask. Clarify that it can't be used with user-provided attention-mask input. I guess the causal-mask is redundant if this input is provided?

File: onnx/defs/nn/defs.cc [hunk 4]
```diff
@@ -3774,7 +3783,21 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+
+          // Add padding mask if kv_nonpad_seqlen is provided
+          if (ctx.hasInput(6)) {
+            builder.Add("KVSeqLenExpanded = Unsqueeze(kv_nonpad_seqlen, One1D)") // [batch_size, 1]
+                .Add("Range = Range(Zero1D, KVSeqLen, One1D)") // [KVSeqLen,]
+                .Add("PaddingMaskBool = Less(Range, KVSeqLenExpanded)") // [batch_size, KVSeqLen]
+                .Add("PaddingMaskFloat = Where(PaddingMaskBool, ScalarZero, FloatInf)") // [batch_size, KVSeqLen]
+                .Add("PaddingMask3D = Unsqueeze(PaddingMaskFloat, One1D)") // [batch_size, 1, KVSeqLen]
+                .Add("PaddingMask4D = Unsqueeze(PaddingMask3D, One1D)") // [batch_size, 1, 1, KVSeqLen]
+                .Add("AttnBiasWithPadding = Add(AttnBias, PaddingMask4D)");
```
Comment (gramalingam) [inline onnx/defs/nn/defs.cc]: Oh ... so this is added to the other mask? I wonder if it has any use-case. Or, if it can cause any issues/problems?

In any case, the other mask is 2D ... so why do we need the unsqueeze to convert it to 4D?

File: onnx/version_converter/convert.h [hunk 5]
```diff
@@ -853,6 +854,7 @@ class DefaultVersionConverter : public BaseVersionConverter {
     registerAdapter(std::make_unique<CompatibleAdapter>("SplitToSequence", OpSetID(23), OpSetID(24)));
 
     /******** 24 -> 23 ********/
+    registerAdapter(std::make_unique<CompatibleAdapter>("Attention", OpSetID(24), OpSetID(23)));
```
Comment (gramalingam) [inline onnx/version_converter/convert.h]: I don't think this is valid. Either we should not register anything ... or we need to validate that it doesn't have the nonpad_kv_selen input.

File: onnx/defs/nn/defs.cc [hunk 6]
```diff
@@ -3774,7 +3793,24 @@ ONNX_OPERATOR_SET_SCHEMA(
                   .Add("AttnBias = ConstantOfShape(AttnBiasShape)");
             }
           }
-          builder.Add("AttnBiasT = Cast (AttnBias)", "to", T1);
+
```
Comment (gramalingam) [inline onnx/defs/nn/defs.cc:3802]: Thanks for updating the documentation to indicate that causal can be used along with attention-mask input. I think the function-body also needs to be updated. It doesn't handle the case where `ctx.hasInput(3)` and `is_causal` are both true ... we need to add both the masks in this case.

File: onnx/defs/nn/defs.cc [hunk 7]
```diff
@@ -3447,6 +3459,13 @@ ONNX_OPERATOR_SET_SCHEMA(
             "past state cache for value with shape `(batch_size, kv_num_heads, past_sequence_length, v_head_size)`",
             "T2",
             OpSchema::Optional)
+        .Input(
+            6,
+            "nonpad_kv_seqlen",
```
Comment (kunal-vaishnavi) [inline onnx/defs/nn/defs.cc:3466]: I've noticed that we often try shorthands in op inputs and then the shorthands haven't scaled well. We tried this in ORT twice with the attention ops: `key_padding_mask` vs `attention_bias` in MHA and `seqlens_k` & `total_sequence_length` vs the new `attention_mask` & `position_ids` inputs in GQA. I'm wondering if it's really worth the effort to have this input vs. just use the `attention_mask` input.

Comment (yuanyao-nv) [reply onnx/defs/nn/defs.cc:3466]: I'm not familiar with the history behind those inputs to the contrib ops you mention. Any particular reason they didn't scale well?
I guess more importantly, any specific concerns with the `nonpad_kv_seqlen` input? It will become important in the case where the K, V inputs are the entire cache buffer, which can contain mostly padded tokens. In such cases, explicitly providing this input allows the backends to skip many unnecessary computations.

File: onnx/defs/nn/defs.cc [hunk 8]
```diff
@@ -3754,44 +3763,46 @@ ONNX_OPERATOR_SET_SCHEMA(
             builder.Add("present_value = Identity (PresentValue)");
           }
 
-          // If attn_mask is provided
+          builder.Add("NewKVSeqLen =  Shape <start = -2, end = -1> (PresentKey)");
+          builder.Add("AttnBiasShape = Concat <axis = -1> (QSeqLen, NewKVSeqLen)");
```
Comment (gramalingam) [inline onnx/defs/nn/defs.cc]: I think we can use `axis = 0`, right? Seems preferable, since it can be interpreted without knowing the rank of input tensors (so downstream tools will likely handle it better).

File: onnx/defs/nn/defs.cc [hunk 9]
```diff
@@ -3341,6 +3341,8 @@ do not contain padded tokens.
 case, the `K` and `V` inputs correspond to the entire cache tensor, so the four optional inputs/outputs past and
 present key and value should not be used. An additional input `nonpad_kv_seqlen` of shape (batch_size,) may be
 provided to indicate the number of non-padding tokens in each sample of the batch to save unnecessary computation.
+Here, the kv_sequence dimension of `attn_mask` can be shorter than `K` and `V`, but still needs to be at least as long
```
Comment (gramalingam) [inline onnx/defs/nn/defs.cc:3344]: If we want to generalize this (how important is it to do this?), then we must also update the documentation of the input, which says it has total-sequence-length. We may need to update the shape-inference method, in case it checks for this.

So, you want to allow smaller, but not bigger?

File: onnx/defs/nn/defs.cc [hunk 10]
```diff
@@ -3443,6 +3445,8 @@ ONNX_OPERATOR_SET_SCHEMA(
             "Shape must be broadcastable to "
             "4D tensor with shape `(batch_size, q_num_heads, q_sequence_length, total_sequence_length)` "
             "where `total_sequence_length = past_sequence_length + kv_sequence_length.` "
+            "The kv_sequence dimension can be smaller than `total_sequence_length` in the case where `K` and `V` are the entire cache tensor "
```
Comment (gramalingam) [inline onnx/defs/nn/defs.cc]: I think it would be better to update the previous two lines also, using a different name than total_sequence_length. But do we really need the generalization? It is likely to confuse people (since we allow smaller, but not larger mask, we allow this in last-dimension, but not others like q_sequence_length). In practice, if users are going to use nonpad_kv_seq_len, they may not use this mask, right?

File: onnx/defs/nn/defs.cc [hunk 11]
```diff
@@ -3754,44 +3763,46 @@ ONNX_OPERATOR_SET_SCHEMA(
...
-            builder.Const1D("ScalarZero", 0.f);
             builder.Add(
-                U == ONNX_NAMESPACE::TensorProto_DataType_BOOL ? "AttnBias = Where(attn_mask, ScalarZero, FloatInf)"
-                                                               : "AttnBias = Identity(attn_mask)");
+                U == ONNX_NAMESPACE::TensorProto_DataType_BOOL ? "AttnBiasShort = Where(attn_mask, ScalarZero, FloatNegInf)"
+                                                               : "AttnBiasShort = Identity(attn_mask)");
+            // If attn_mask has a shorter kv sequence length, we pad it to NewKVSeqLen with FloatNegInf
+            builder.Add("MaskKVSeqLen = Shape <start = -1> (attn_mask)")
+                .Add("PaddingKVSeqLen = Sub(NewKVSeqLen, MaskKVSeqLen)")
+                .Add("MaskPrefixShape = Shape <start = 0, end = -1> (attn_mask)")
```
Comment (gramalingam) [inline onnx/defs/nn/defs.cc]: minor nit: may be a bit shorter to use Pad. But it's not a big deal. Not sure I understand the Cast down below, it seems redundant? Aren't we using float until the very last point, and casting to target type then?

File: onnx/defs/nn/defs.cc (hunk 9 above)
Comment (yuanyao-nv) [reply onnx/defs/nn/defs.cc:3344]: @gramalingam 
It doesn't seem like shape inference uses this attn_mask input at all. So that part can probably stay as is.
What are you referring to by "generalize"?
The kv_seqlen dimension can be smaller than the K/V inputs's seqlen dimension in use case 2 of kv cache. There should not be any reason for it to be bigger than K/V. Maybe I should update the description of this input to explicitly state the dimensions for the two cases of kv cache usage.

File: onnx/defs/nn/defs.cc (hunk 11 above)
Comment (yuanyao-nv) [reply onnx/defs/nn/defs.cc]: The reason for the CastLike is that the input mask can come in a type that is not FP32, but `MaskPadding` is created in FP32. So we need to cast it before Concat.

File: onnx/defs/nn/defs.cc (hunk 1 above)
Comment (yuanyao-nv) [reply onnx/defs/nn/defs.cc]: Updated to "nonpad_kv_seqlen" for now.

File: onnx/defs/nn/defs.cc (hunk 3 above)
Comment (yuanyao-nv) [reply onnx/defs/nn/defs.cc]: Added a description of the two KV cache use cases.
There's nothing preventing `nonpad_kv_seqlen` from being used with either is_causal and/or attn_mask though.

File: onnx/version_converter/convert.h (hunk 5 above)
Comment (yuanyao-nv) [reply onnx/version_converter/convert.h]: I added a 24->23 converter to explicitly check that the nonpad_ke_seqlen input is not there.

File: onnx/defs/nn/defs.cc (hunk 8 above)
Comment (yuanyao-nv) [reply onnx/defs/nn/defs.cc]: Right. Updated.

File: onnx/defs/nn/defs.cc (hunk 11 above)
Comment (yuanyao-nv) [reply onnx/defs/nn/defs.cc]: Updated to use Pad instead.

File: onnx/defs/nn/defs.cc (hunk 10 above)
Comment (yuanyao-nv) [reply onnx/defs/nn/defs.cc]: I kind of already answered briefly in the other thread.
Allowing smaller mask makes sense if the K V inputs have mostly padding tokens. There's no reason to allow bigger. 
For Q padding, masking doesn't really work because of the softmax and it's usually handled by the runtime instead. So there's not a reason to allow mask to be shorter in q_seqlen dimension.
Is_causal, attn_mask, nonpad_kv_seqlen are all additive masks, so there should not be anything preventing them from being used together.

Comment (gramalingam) [reply onnx/defs/nn/defs.cc]: Ok. I suggest changing `total_sequence_length` in previous two lines to some other word: may be `attn_sequence_length` or something like that. Then, the documentation can say that `attn_sequence_length` must be <= total_sequence_length, and that it will be padded to total_sequence_length with minus infinity values if it is smaller.

**Review by gramalingam — APPROVED** [review]
LGTM, thanks ... just a couple of minor comments left about documentation of attn_sequence_length

File: onnx/defs/nn/defs.cc (hunk 10 above)
Comment (yuanyao-nv) [reply onnx/defs/nn/defs.cc]: Actually I think we can just say this
```
            "Shape must be broadcastable to `(batch_size, q_num_heads, q_sequence_length, total_sequence_length)` "
            "where `total_sequence_length = past_sequence_length + kv_sequence_length.` "
            "The last dimension can also be shorter than `total_sequence_length` and will be padded to `total_sequence_length` with negative infinity, "
            "but it still needs to be at least as long as the maximum value of `nonpad_kv_seqlen` if provided. "
```
This should cover all three cases: 
1. no kv cache at all 
2. kv cache concat happens inside Attention op with past/present_key/value I/O
3. K/V inputs are entire cache buffer

Comment (gramalingam) [reply onnx/defs/nn/defs.cc]: That's fine. I still have one quibble about the last line "but it still needs to be at least as long as the maximum ..." ... I don't think we are checking for this condition in the shape-inference, are we? If we are not checking for this condition, it would be simpler to drop this condition here ... because padding to the required size can be done no matter what its length is, as long as it is smaller. And if this works for the user, then that's fine.

File: onnx/defs/nn/defs.cc [hunk 12]
```diff
@@ -3735,46 +3760,64 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+            builder.Add("AttnBiasCausal = Identity(AttnBias)");
+          }
+
+          // Add padding mask if kv_nonpad_seqlen is provided
+          if (ctx.hasInput(6)) {
+            if (!is_3d_input) {
+              builder.Add("KVSeqLen = Shape <start = -2, end = -1> (K)");
             }
+            builder
+                .Add("KVSeqLenExpanded = Unsqueeze(nonpad_kv_seqlen, One1D)")  // [batch_size, 1]
```
Comment (yuanyao-nv) [inline onnx/defs/nn/defs.cc]: @justinchuby Hmm any thoughts why the linter is complaining about these lines? https://github.com/onnx/onnx/actions/runs/16608616499/job/46986673953?pr=7164
I cannot repro this locally. What the CI linter sees also seems to be a bit different from what's written here (0 space between code and comment).

File: onnx/defs/nn/defs.cc (hunk 10 above)
Comment (yuanyao-nv) [reply onnx/defs/nn/defs.cc]: Yea that's right, we are not checking this condition in shape inference. Let me drop this sentence then.

File: onnx/defs/nn/defs.cc (hunk 12 above)
Comment (justinchuby) [reply onnx/defs/nn/defs.cc]: I think it is asking for a single space before the comments

Comment (yuanyao-nv) [reply onnx/defs/nn/defs.cc]: That does seem to be the reason. thanks! although in the optional lint there are complaints like "At least two spaces is best between code and comments  [whitespace/comments]"

File: onnx/defs/nn/defs.cc [hunk 13]
```diff
@@ -3735,46 +3759,64 @@ ONNX_OPERATOR_SET_SCHEMA(
...
+          }
+
+          // Add padding mask if kv_nonpad_seqlen is provided
+          if (ctx.hasInput(6)) {
+            if (!is_3d_input) {
+              builder.Add("KVSeqLen = Shape <start = -2, end = -1> (K)");
             }
+            builder
+                .Add("KVSeqLenExpanded = Unsqueeze(nonpad_kv_seqlen, One1D)") // [batch_size, 1]
+                .Add("Range = Range(Zero1D, KVSeqLen, One1D)") // [KVSeqLen,]
```
Comment (titaiwangms) [inline onnx/defs/nn/defs.cc:3810]: The inputs should be Scalar: https://github.com/onnx/onnx/blob/main/docs/Operators.md#Range

cc @gramalingam @yuanyao-nv @justinchuby

Comment (titaiwangms) [reply onnx/defs/nn/defs.cc:3810]: It's caught by ORT, but ONNX checker does not complain about this for some reasons..

Comment (titaiwangms) [reply onnx/defs/nn/defs.cc:3810]: Just like RMSNorm: https://github.com/onnx/onnx/pull/7135/files (reference fix)

Comment (yuanyao-nv) [reply onnx/defs/nn/defs.cc:3810]: @titaiwangms how about this?
```
                .Const("Zero0D", (int64_t)(0))
                .Const("One0D", (int64_t)(1))
                .Add("KVSeqLen0D = Unsqueeze(KVSeqLen, Zero1D)")
                .Add("Range = Range(Zero0D, KVSeqLen0D, One0D)")
```

Comment (justinchuby) [reply onnx/defs/nn/defs.cc:3810]: Thanks for catching this. We missed it

Comment (yuanyao-nv) [reply onnx/defs/nn/defs.cc:3810]: PR to fix: https://github.com/onnx/onnx/pull/7240
